### ISSUE TITLE
Use require() type module-loading to facilitate unit-testing, lazy-loading and interoperability with NPM

### ISSUE THREAD

**Issue by sirockin** [json-editor/json-editor#447]
# General Information

The current build approach (using Grunt to concatenate files) is simple and transparent but makes it very difficult to export code for unit-testing. It also means that in production, all code gets loaded at once, whether it is a core feature or one that is only used by a minority of users, potentially increasing load times.

# References

* #60 
* #428 

# Proposed Solution(s)

There are several build tools available to allow require()-style imports - the most popular these days seem to be:

* [Browserify](http://browserify.org/index.html)
* [Webpack](https://webpack.js.org/). 

[This article](https://www.cleveroad.com/blog/gulp-browserify-webpack-grunt) gives a pretty good overview of their pros and cons.  

tldr: Browserify provides just the require() import facility, Webpack provides a whole lot of other features, allowing you to get rid of Grunt altogether and is the standard tool installed with Angular.

# Conclusion

Implementing one of the above approaches for the repo would be a few hours' work. I'm not an expert but will try to find the time to create a PR if there is a reasonable chance it would be accepted. Before doing so though, I'd like a steer from contributors about which technology you'd prefer to adopt.

**Comment by schmunk42** [comment]
> if there is a reasonable chance it would be accepted

Yes, we would go into that direction :)

**Comment by pmk65** [comment]
@sirockin Please use the 2.x/develop branch if you are going to submit a PR.
From reading the article, Webpack seems to be the logic choice, as it offers more features than Browserify.

**Comment by sirockin** [comment]
@pmk65 from reading that article Webpack would be my preferred direction too. In that case would you be happy to lose Grunt? It seems fairly redundant in this situation.

**Comment by pmk65** [comment]
If same/similar operations can be performed by Webpack, I see no problem in abandoning Grunt.
